src/client/build/register.ts

```diff
@@ -0,0 +1,83 @@
+import { Workbox, messageSW } from 'workbox-window'
+import { RegisterSWOptions } from '../type'
+
+// __SW_AUTO_UPDATE__ will be replaced by virtual module
+const autoUpdateMode = '__SW_AUTO_UPDATE__'
+
+// @ts-ignore
+const auto = autoUpdateMode === 'true'
+
+export type { RegisterSWOptions }
+
+export function registerSW(options: RegisterSWOptions = {}) {
+  const {
+    immediate = false,
+    onNeedRefresh,
+    onOfflineReady,
+  } = options
+
+  let wb: Workbox | undefined
+  let registration: ServiceWorkerRegistration | undefined
+
+  const updateServiceWorker = async(reloadPage = true) => {
+    if (!auto) {
+      // Assuming the user accepted the update, set up a listener
+      // that will reload the page as soon as the previously waiting
+      // service worker has taken control.
+      if (reloadPage) {
+        wb?.addEventListener('controlling', (event) => {
+          if (event.isUpdate)
+            window.location.reload()
+        })
+      }
+
+      if (registration && registration.waiting) {
+        // Send a message to the waiting service worker,
+        // instructing it to activate.
+        // Note: for this to work, you have to add a message
+        // listener in your service worker. See below.
+        await messageSW(registration.waiting, { type: 'SKIP_WAITING' })
+      }
+    }
+  }
+
+  if ('serviceWorker' in navigator) {
+    // __SW__ and __SCOPE__ will be replaced by virtual module
+    wb = new Workbox('__SW__', { scope: '__SCOPE__' })
+
+    wb.addEventListener('activated', (event) => {
+      // this will only controls the offline request.
+      // event.isUpdate will be true if another version of the service
+      // worker was controlling the page when this version was registered.
+      if (event.isUpdate)
+        auto && window.location.reload()
+      else
+        onOfflineReady?.()
+    })
+
+    if (!auto) {
```

Comment:
If the application is not going to respond to the event, does it make sense that we register the listeners when `injectRegister: 'auto'` and `registerType: 'prompt'` ? 

Should we throw an Error if `onNeedRefresh` not provided or just omit adding event listeners?

```javascript
if (!auto) {
  if (!onNeedRefresh) throw new Error('You need to provide onNeedRefresh on registerSW options')
}
```
or
```javascript
if (!auto && onNeedRefresh) {
```

Reply:
or
```javascript
if (!auto) {
  if (!onNeedRefresh) console.warn('VitePWA: You need to provide onNeedRefresh on registerSW options')
}
```
or
```javascript
if (!auto) {
  if (!onNeedRefresh) console.error('VitePWA: You need to provide onNeedRefresh on registerSW options')
}
```

Reply:
I don't think we would need to check it or make it required, we just document the usage, and its users' choose to use this callback or ignore it.

Reply:
and how about to just warn the user? you forgot that now it is the default configuration